File: epi_judge_js/buyAndSellStock.js
const assert = require('assert');

function buyAndSellStock(prices) {
    let minPriceSoFar = Infinity;
    let maxProfit = 0;

    prices.forEach((price) => {
        let maxProfitToday = price - minPriceSoFar;
        maxProfit = Math.max(maxProfit, maxProfitToday);
        minPriceSoFar = Math.min(minPriceSoFar, price);
    });

    return maxProfit;
}

function buyAndSellStockv2(prices) {
    let minPriceSoFar;
    let maxProfit = 0;

    prices.forEach((price, idx) => {
        if (idx === 0) {
            minPriceSoFar = price;
        } else {
            minPriceSoFar = Math.min(minPriceSoFar, price);
            let profitToday = price - minPriceSoFar;
            maxProfit = Math.max(maxProfit, profitToday); 
        }
    })

    return maxProfit;
}

const prices = [310, 315, 275,295, 260, 270, 290, 230, 255, 250];
const ans = 30;

assert.strictEqual(buyAndSellStockv2(prices), ans);

File: epi_judge_python/buy_and_sell_stock.py
from typing import List

from test_framework import generic_test


def buy_and_sell_stock_once(prices: List[float]) -> float:
    min_price_so_far = float('inf')
    max_profit = 0

    for price in prices:
        max_profit_today = price - min_price_so_far
        min_price_so_far = min(price, min_price_so_far)
        max_profit = max(max_profit, max_profit_today)

    return max_profit

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('buy_and_sell_stock.py',
                                       'buy_and_sell_stock.tsv',
                                       buy_and_sell_stock_once))
